File: src/external/git.rs
use std::io::{IsTerminal, Write};
use std::path::Path;
use std::process::{Command, Stdio};

use anyhow::Result;

use crate::common::log::LogResult;

pub fn commit<S>(message: S) -> Result<()>
where
    S: AsRef<str>,
{
    let mut cmd = Command::new("git");
    cmd.args(["commit", "--file=-"])
        .stdin(Stdio::piped())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit());
    let mut child = cmd.spawn().log()?;
    child
        .stdin
        .as_ref()
        .unwrap()
        .write_all(message.as_ref().as_bytes())
        .log()?;
    let status = child.wait().log()?;
    crate::ensure!(status.success());
    Ok(())
}

pub fn diff<I, J, S, T>(exclude: I, include: J) -> Result<String>
where
    I: IntoIterator<Item = S>,
    J: IntoIterator<Item = T>,
    S: AsRef<Path>,
    T: AsRef<Path>,
{
    let mut cmd = Command::new("git");
    cmd.args(["diff", "--cached"]);
    exclude.into_iter().for_each(|p| {
        cmd.arg(format!(":(exclude){}", p.as_ref().to_str().unwrap()));
    });
    include.into_iter().for_each(|p| {
        cmd.arg(p.as_ref().to_str().unwrap());
    });
    cmd.stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit());
    let output = cmd.output().log()?;
    crate::ensure!(output.status.success());
    String::from_utf8(output.stdout).log()
}

pub fn status<I, J, S, T>(exclude: I, include: J) -> Result<()>
where
    I: IntoIterator<Item = S>,
    J: IntoIterator<Item = T>,
    S: AsRef<Path>,
    T: AsRef<Path>,
{
    let mut cmd = Command::new("git");
    if std::io::stdout().is_terminal() {
        cmd.args(["-c", "color.status=always"]);
    }
    cmd.args(["status"]);
    exclude.into_iter().for_each(|p| {
        cmd.arg(format!(":(exclude){}", p.as_ref().to_str().unwrap()));
    });
    include.into_iter().for_each(|p| {
        cmd.arg(p.as_ref().to_str().unwrap());
    });
    cmd.stdin(Stdio::null())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit());
    let status = cmd.status().log()?;
    crate::ensure!(status.success());
    Ok(())
}

File: ai_commit_cli/external/git.py
import subprocess
import sys
from collections.abc import Sequence


def commit(message: str) -> None:
    subprocess.run(
        args=["git", "commit", "--file=-"],
        stdout=sys.stdout,
        stderr=sys.stderr,
        check=True,
        input=message,
        text=True,
    )


def diff(exclude: Sequence[str], include: Sequence[str]) -> str:
    completed: subprocess.CompletedProcess = subprocess.run(
        args=[
            "git",
            "diff",
            "--cached",
            *[":(exclude)" + p for p in exclude],
            *include,
        ],
        stdin=subprocess.DEVNULL,
        stdout=subprocess.PIPE,
        stderr=sys.stderr,
        check=True,
        text=True,
    )
    return completed.stdout


def status(exclude: Sequence[str], include: Sequence[str]) -> str:
    completed: subprocess.CompletedProcess = subprocess.run(
        args=[
            "git",
            *(["-c", "color.status=always"] if sys.stdout.isatty() else []),
            "status",
            *[":(exclude)" + p for p in exclude],
            *include,
        ],
        stdin=subprocess.DEVNULL,
        stdout=sys.stdout,
        stderr=sys.stderr,
        check=True,
        text=True,
    )
    return completed.stdout
